xpasses/XS002/XS002.go

```diff
@@ -0,0 +1,51 @@
+package XS002
+
+import (
+	"go/ast"
+	"sort"
+
+	"golang.org/x/tools/go/analysis"
+
+	"github.com/bflad/tfproviderlint/helper/terraformtype/helper/schema"
+	"github.com/bflad/tfproviderlint/passes/commentignore"
+	"github.com/bflad/tfproviderlint/passes/helper/schema/schemamapcompositelit"
+)
+
+const Doc = `check for Schema that attribute names are in alphabetical order
+
+The XS002 analyzer reports cases of schemas where attribute names
+are not in alphabetical order.`
+
+const analyzerName = "XS002"
+
+var Analyzer = &analysis.Analyzer{
+	Name: analyzerName,
+	Doc:  Doc,
+	Requires: []*analysis.Analyzer{
+		schemamapcompositelit.Analyzer,
+		commentignore.Analyzer,
+	},
+	Run: run,
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	ignorer := pass.ResultOf[commentignore.Analyzer].(*commentignore.Ignorer)
+	schemamapcompositelits := pass.ResultOf[schemamapcompositelit.Analyzer].([]*ast.CompositeLit)
+
+	for _, smap := range schemamapcompositelits {
+		if ignorer.ShouldIgnore(analyzerName, smap) {
+			continue
+		}
+
+		schemaKeys := make([]string, 0 , len(schema.GetSchemaMapAttributeNames(smap)))
+		for _, attributeName := range schema.GetSchemaMapAttributeNames(smap) {
```

Comment:
hmm I may have missed this...is there a helper method available that already returns the array of (*ast.BasicLit).Value rather than having to loop through here?

Reply:
The AST map keys are not guaranteed to be of the `ast.BasicLit` type (e.g. `"stringkey"`), but instead may be other AST types, such as those for a constant/variable reference (e.g. `myAwesomeConstKey`). We'll want to use `astutils.ExprStringValue(attributeName)` which returns a `*string` for type safety in this case.

```go
for _, attributeName := range schema.GetSchemaMapAttributeNames(smap) {
	if v := astutils.ExprStringValue(attributeName); v != nil {
		schemaKeys = append(schemaKeys, *v)
	}
}
```

Potentially wrapped up in a helper near the other one, say

```go
// GetSchemaMapAttributeNameStrings returns schema map keys that are literal strings
//
// This function does not handle constant, variable, or other types of keys.
func GetSchemaMapAttributeNameStrings(compositeLit []*ast.CompositeLit) []string
```

We may want to add testing that ensures this code doesn't panic with other types of keys like:

```go
const attrKey = "attr_key"

map[string]*schema.Schema{
	"arn": {},
	"name": {},
	attrKey: {},
}
```